### PR TITLE
[batch] maybe fix cp dirs

### PR DESCRIPTION
**Pull request by danking**

### CODE REVIEW

**Review by jigold — CHANGES_REQUESTED** [review]

File: batch/batch/worker/worker.py [hunk 1]
```diff
@@ -542,16 +542,29 @@ def copy_command(src, dst, requester_pays_project=None):
...
         requester_pays_project = ''
 
+    def gsutil_cp_r(src, dst, * recursive):
+        flags = '-R' if recursive else ''
+        return f'retry gsutil {requester_pays_project} -m cp {flags} {shq(src)} {shq(dst)}'
+
     if not dst.startswith('gs://'):
-        mkdirs_file = f'mkdir -p {shq(os.path.dirname(dst))} && '
-        mkdirs_dir = f'mkdir -p {shq(dst)} && '
+        target_directory = os.path.dirname(dst)
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/worker/worker.py:550]: This is specifically for the source. I'm worried about `gs://hail-jigold/`

[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/worker/worker.py:550]: Maybe not, actually. Forget this comment.

File: batch/batch/worker/worker.py [hunk 2]
```diff
@@ -542,16 +542,29 @@ def copy_command(src, dst, requester_pays_project=None):
...
         requester_pays_project = ''
 
+    def gsutil_cp_r(src, dst, *, recursive):
+        flags = '-R' if recursive else ''
+        return f'retry gsutil {requester_pays_project} -m cp {flags} {shq(src)} {shq(dst)}'
+
     if not dst.startswith('gs://'):
-        mkdirs_file = f'mkdir -p {shq(os.path.dirname(dst))} && '
-        mkdirs_dir = f'mkdir -p {shq(dst)} && '
+        target_directory = os.path.dirname(dst)
```
Comment (jigold) [inline batch/batch/worker/worker.py:550]: Maybe this needs to be `os.path.dirname(dst.rstrip('/') + '/')`?

File: batch/batch/worker/worker.py [hunk 3]
```diff
@@ -542,16 +542,29 @@ def copy_command(src, dst, requester_pays_project=None):
...
+        target_basename = os.path.basename(dst)
```
Comment (jigold) [inline batch/batch/worker/worker.py:551]: Same here. We are assuming that dst is a directory.

**Review by jigold — DISMISSED** [review]

File: batch/batch/worker/worker.py [hunk 4]
```diff
@@ -537,21 +537,35 @@ async def add_gcsfuse_bucket(mount_path, bucket, key_file, read_only):
 
 
 def copy_command(src, dst, requester_pays_project=None):
+    src = src.rstrip('/')  # ensure that the source_basename is always non-empty
```
Comment (danking) [inline batch/batch/worker/worker.py:540]: @jigold this was the bug. I believe it will work now.

**Review by jigold — APPROVED** [review]

**Pull request merged**